src/main/scala/com/sumologic/sumobot/http_frontend/authentication/AlbJwtAuthentication.scala

```diff
@@ -0,0 +1,127 @@
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.util.Success
+
+object AlbJwtAuthentication {
+  private val HeaderName = "x-amzn-oidc-data"
+  private val FailureResponse = AuthenticationForbidden(HttpResponse(401))
+  private val KeyEndpointTimeout = 5.seconds
+
+  private val KeyHeaderBegin = "-----BEGIN PUBLIC KEY-----"
+  private val KeyHeaderEnd = "-----END PUBLIC KEY-----"
+
+  private val LogoutEndpoint = "/logout"
+  private val LogoutEndpointLink = "logout"
+}
+
+class AlbJwtAuthentication(config: Config) extends HttpAuthentication {
+  implicit val system = ActorSystem()
+  implicit val materializer = ActorMaterializer()
+
+  private val keyEndpoint = config.getString("key-endpoint")
+  private val cookieName = config.getString("cookie-name")
+  private val logoutRedirect = config.getString("logout-redirect")
+
+  override def authentication(request: HttpRequest): AuthenticationResult = {
+    request.headers.find(header => header.is(HeaderName)) match {
+      case Some(header) => parseJwtString(header.value())
+      case _ => FailureResponse
+    }
+  }
+
+  private val redirectResponse = HttpResponse(StatusCodes.TemporaryRedirect, headers = List(`Location`(logoutRedirect),
+    RawHeader("Set-Cookie", s"$cookieName=; Expires=-1; Path=/; Secure; HttpOnly")))
+
+  override def routes: PartialFunction[HttpRequest, HttpResponse] = {
+    case req@HttpRequest(GET, Uri.Path(LogoutEndpoint), _, _, _) =>
+      redirectResponse
+
+    case req@HttpRequest(HEAD, Uri.Path(LogoutEndpoint), _, _, _) =>
+      redirectResponse
+
+    case req@HttpRequest(OPTIONS, Uri.Path(LogoutEndpoint), _, _, _) =>
+      HttpResponse()
+        .withHeaders(List(`Access-Control-Allow-Methods`(List(GET))))
+  }
+
+  private def parseJwtString(jwtString: String): AuthenticationResult = {
+    JwtJson.decodeAll(jwtString, JwtOptions(signature = false)) match {
+      case Success((header, claim, _)) =>
+        downloadKey(header).map {
+          key =>
+            val publicKey = parseKey(key)
+
+            if (JwtJson.isValid(jwtString, publicKey, Seq(JwtAlgorithm.ES256))) {
+              val parsedClaim = Json.parse(claim.content)
+              val nameOption = (parsedClaim \\ "name").headOption
+              nameOption match {
+                case Some(name) =>
+                  val authMessage = Some(s"Logged in as: ${name.as[String]}")
+                  val links = Seq(Link("Log out", LogoutEndpointLink))
+                  AuthenticationSucceeded(AuthenticationInfo(authMessage, links))
+                case None => FailureResponse
+              }
+            } else {
+              FailureResponse
+            }
+        }.getOrElse(FailureResponse)
+      case _ => FailureResponse
+    }
+  }
+
+  private def downloadKey(header: JwtHeader): Option[String] = {
+    header.keyId match {
+      case Some(keyId) =>
+        val request = HttpRequest(uri = s"$keyEndpoint$keyId")
```

Comment:
Are you downloading a key on every request?  This seems excessive and adds a lot of latency on each request. Maybe we can cache this key in memory?

Reply:
Measuring using Chrome, the this part takes ~180ms (maybe less). 

In normal use-case, browser makes only 2 requests (HTTP + WebSocket) and then the user spends a longer time using already open WebSocket connection (no futher auth).

If I were to cache it, what caching strategy would you suggest (fixed sized buffer? queue?something more advanced?) (is this additional complexity really needed?)

Reply:
Update: regarding caching: it's probably easy to do - some Akka HTTP caching options to set.